### PR TITLE
[Application] Implement MsgUnstakeApplication & Add Extensive Tests

### PR DESCRIPTION
**Pull request by Olshansk**
### Human Summary

- Implemented MsgApplicationUnstake
- Added the necessary tests, comments & TODOs
- Using the newer SDK errors module across the board

https://github.com/pokt-network/poktroll/assets/1892194/c707aad2-2260-49cb-ba99-f369b4fb45b2

### Summary generated by Reviewpad on 18 Oct 23 19:19 UTC
This pull request includes changes to multiple files. 

In the file "MsgUnstakeApplication.go":
- An option named "(cosmos.msg.v1.signer)" with the value "address" has been added to the "MsgUnstakeApplication" message. More information about this option can be found at https://docs.cosmos.network/main/build/building-modules/messages-and-queries.

In the file "errors.go":
- The import path "github.com/cosmos/cosmos-sdk/types/errors" has been updated to "cosmossdk.io/errors".

In the file "message_unstake_application_test.go":
- Removed the import of "github.com/cosmos/cosmos-sdk/types/errors".
- Added the import of "pocket/testutil/sample".
- Added test cases for a valid "MsgUnstakeApplication" and invalid "MsgUnstakeApplication" with missing or invalid address.

In the file "message_stake_gateway.go":
- The import "cosmossdk.io/errors" has been renamed to "sdkerrors".
- The variable name "stakeAmount" has been changed to "stake".
- Error messages in "errors.Wrapf" have been updated to include the new variable name "stake".
- Validation of "stakeAmount" has been updated to use the "IsValid" method instead of "Validate".
- Variable name "stakeAmount" has been changed to "stake".
- Variable name "stakeAmount" has been changed to "stake".

In the file "message_unstake_supplier.go":
- The import statement "cosmossdk.io/errors" has been changed to "sdkerrors "cosmossdk.io/errors"".
- The function "ValidateBasic()" now returns an error of type "sdkerrors" instead of "errors".
- The error message in the return statement has been changed to use "sdkerrors.Wrapf" instead of "errors.Wrapf".

In the file "message_unstake_gateway.go":
- The import statement "cosmossdk.io/errors" has been changed to "sdkerrors "cosmossdk.io/errors"".
- The function "ValidateBasic()" now returns an error of type "sdkerrors" instead of "errors".
- The error message in the return statement has been changed to use "sdkerrors.Wrapf" instead of "errors.Wrapf".

In the file "tx_unstake_application.go":
- The command "CmdUnstakeApplication()" now includes additional information about unstaking an application.
- The "Long" description explains how to unstake an application and provides an example.
- The "Args" field in the command now accepts exactly 0 arguments.
- The "RunE" function has been modified to use the updated "client.GetClientTxContext()" method.

In the file "x/gateway/types/message_stake_gateway.go":
- The import "cosmossdk.io/errors" has been replaced with "sdkerrors" as the import path.
- The variable "stakeAmount" has been renamed to "stake" and is now of type "sdk.Coin".
- The "stakeAmount" variable is now validated using "stake.IsValid()" instead of "stakeAmount.IsValid()".
- The variable "stakeAmount" has been replaced with "stake" in all other references.
- The check for a stake amount being zero or negative is now done using "stake.IsZero()" and "stake.IsNegative()" instead of "stakeAmount.IsZero()" and "stakeAmount.IsNegative()".
- The check for the "stakeAmount" denom being equal to "upokt" is now done using "stake.Denom" instead of "stakeAmount.Denom".
- The "ErrGatewayInvalidAddress" and "ErrGatewayInvalidStake" errors have been changed to "sdkerrors.Wrapf" functions.

In the file "expected_keepers.go":
- The BankKeeper interface has a new method called "UndelegateCoinsFromModuleToAccount" which allows for the undelegation of coins from a module to an account.

Please review these changes and ensure they are appropriate for the pull request.

## Issue

- #8

## Type of change

Select one or more:

- [x] New feature, functionality or library
- [ ] Bug fix
- [ ] Code health or cleanup
- [ ] Documentation
- [ ] Other (specify)

## Testing

- [x] **Run all unit tests**: `make go_test`
- [ ] **Verify Localnet manually**: See the instructions [here](TODO: add link to instructions)

## Sanity Checklist

- [x] I have tested my changes using the available tooling
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, updated documentation and left TODOs throughout the codebase

### CODE REVIEW

File: x/gateway/types/message_stake_gateway.go [hunk 1]
```diff
@@ -39,7 +39,8 @@ func (msg *MsgStakeGateway) GetSignBytes() []byte {
 func (msg *MsgStakeGateway) ValidateBasic() error {
 	_, err := sdk.AccAddressFromBech32(msg.Address)
 	if err != nil {
-		return sdkerrors.Wrapf(sdkerrors.ErrInvalidAddress, "invalid address address (%s)", err)
+		// TODO(@h5law): Replace with a proper error
```
Comment (h5law) [inline x/gateway/types/message_stake_gateway.go:42]: Merge with main for this to be solved

File: x/application/keeper/msg_server_unstake_application_test.go [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+	wctx := sdk.WrapSDKContext(ctx)
+
+	// Generate an address for the application
+	addr := sample.AccAddress()
+
+	// Verify that the app does not exist yet
+	_, isAppFound := k.GetApplication(ctx, addr)
+	require.False(t, isAppFound)
+
+	// Prepare the application
```
Comment (h5law) [inline x/application/keeper/msg_server_unstake_application_test.go:27]:
```suggestion
	// Prepare the application
	initialStake := sdk.NewCoin("upokt", sdk.NewInt(100))
```

File: x/application/keeper/msg_server_unstake_application_test.go [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+
+	// Stake the application
+	_, err := srv.StakeApplication(wctx, app)
+	require.NoError(t, err)
+
+	// Verify that the application exists
+	foundApp, isAppFound := k.GetApplication(ctx, addr)
+	require.True(t, isAppFound)
+	require.Equal(t, addr, foundApp.Address)
+	require.Equal(t, int64(100), foundApp.Stake.Amount.Int64())
```
Comment (h5law) [inline x/application/keeper/msg_server_unstake_application_test.go:41]:
```suggestion
	require.Equal(t, initialStake.Amount, foundApp.Stake.Amount)
```

File: x/application/types/message_unstake_application.go [hunk 4]
```diff
@@ -1,8 +1,8 @@
 package types
 
 import (
+	sdkerrors "cosmossdk.io/errors"
```
Comment (h5law) [inline x/application/types/message_unstake_application.go:4]: NIT: Not major but in places we import this as `errorsmod` and others `sdkerrors` I am not aliasing it and just calling it `errors` (we dont use native `errors` package as this does wrapping for us so basically multierr too).

Just think consistency here would be better but not a huge deal

File: x/gateway/genesis.go [hunk 5]
```diff
@@ -2,15 +2,16 @@ package gateway
...
 	"pocket/x/gateway/keeper"
 	"pocket/x/gateway/types"
 )
 
 // InitGenesis initializes the module's state from a provided genesis state.
 func InitGenesis(ctx sdk.Context, k keeper.Keeper, genState types.GenesisState) {
 	// Set all the gateway
-	for _, elem := range genState.GatewayList {
-		k.SetGateway(ctx, elem)
+	for _, gateway := range genState.GatewayList {
```
Comment (h5law) [inline x/gateway/genesis.go:13]: Addressed in #75 maybe worth undoing these to avoid conflicts

File: x/gateway/types/genesis.go [hunk 6]
```diff
@@ -22,29 +22,29 @@ func DefaultGenesis() *GenesisState {
 func (gs GenesisState) Validate() error {
 	gatewayIndexMap := make(map[string]struct{})
 
-	for _, elem := range gs.GatewayList {
+	for _, gateway := range gs.GatewayList {
```
Comment (h5law) [inline x/gateway/types/genesis.go:25]: See https://github.com/pokt-network/poktroll/pull/72/files#r1364288032 think better to leave this to avoid conflicts

File: x/gateway/types/message_stake_gateway.go [hunk 7]
```diff
@@ -42,7 +43,7 @@ func (msg *MsgStakeGateway) ValidateBasic() error {
 	// Validate the address
 	_, err := sdk.AccAddressFromBech32(msg.Address)
 	if err != nil {
-		return errors.Wrapf(ErrGatewayInvalidAddress, "invalid gateway address %s; (%v)", msg.Address, err)
+		return sdkerrors.Wrapf(ErrGatewayInvalidAddress, "invalid gateway address %s; (%v)", msg.Address, err)
```
Comment (h5law) [inline x/gateway/types/message_stake_gateway.go:46]: Here we are double importing `cosmossdk.io/errors` I am not aliasing so this doesnt change anything except adding a double import see: https://github.com/pokt-network/poktroll/pull/72/files#r1364287010 for a NIT around import naming convenction on this package I think we should come to aggreement on

File: x/gateway/types/message_unstake_gateway.go [hunk 8]
```diff
@@ -39,7 +39,7 @@ func (msg *MsgUnstakeGateway) GetSignBytes() []byte {
 func (msg *MsgUnstakeGateway) ValidateBasic() error {
 	_, err := sdk.AccAddressFromBech32(msg.Address)
 	if err != nil {
-		return sdkerrors.Wrapf(sdkerrors.ErrInvalidAddress, "invalid address address (%s)", err)
+		return sdkerrors.Wrapf(ErrGatewayInvalidAddress, "invalid address address (%s)", err)
```
Comment (h5law) [inline x/gateway/types/message_unstake_gateway.go:42]: I would avoid these changes as they are addressed in #75 as these are from the scaffolding alone addressing them here seems out of scope of this PR

**Review by h5law — CHANGES_REQUESTED** [review]
Overall PR looks great, left some comments around things I have covered in the Gateway version of this PR. Other than that I think the only thing outstanding is naming of the errors import we are using as this is sometimes `sdkerrors` `errorsmod` or as I use just `errors` (no alias).

Besides that great work! The previous PR laid great foundations and this is just continuing on that train. Love it

File: x/gateway/types/message_unstake_gateway.go (hunk 8 above)
Comment (Olshansk) [reply x/gateway/types/message_unstake_gateway.go:42]: I already have it in place and I think doing a `git merge` + conflict resolution would be easier than checking out all the places I need to revert.

File: x/gateway/genesis.go (hunk 5 above)
Comment (Olshansk) [reply x/gateway/genesis.go:13]: I think conflict resolution with a `git merge` will be quite easy.

If this is something you find yourself spending a lot on time on (i.e. > 2 minutes), let me know and I can show you quick ways of getting it done.

**Review by h5law — APPROVED** [review]
LGTM LFGGGGGG :shipit:
